# Introduction
- In this project, we'll work with a datset os used cars from eBay Kleinanzeigen, a [classified](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay websites

- The dataset was originally scraped & uploaded to Kaggle by user orgeslaka. However, it's not available anymore, but you can find it [here](https://data.world/data-society/used-cars-data)

### Goals:
- To clean the data & analyze the included used car listings

### Some modifications to notice
- The dataset we're working below only contained 50,000 sample points from the full one (from [**Dataquest**](https://www.dataquest.io/) and it has been modified a bit to make it more closely resemble what we'll expect from a scraped dataset)

## Data dictionary:
`dateCrawled` - when this as was first crawled. All field-values are taken from this date

`name` - name of the car

`seller` - whether the seller is private or a dealer

`offerType` - the Type of listing

`price` - the price on the ad to sell the car

`abtest` - whether the listing is included in an A/B test

`vehicleType` - the vehicle Type

`yearOfRegistration` - the year in which the car was first registered

`powerPS` - the power of the car in PS

`model` - the car model name

`kilometer` - how many kilometers the car has driven

`monthOfRegistration` - the month in which the car was first registerd

`fuelType` - what type of fuel the car uses

`brand` - the brand of the car

`notRepairedDamage` - If the car has a damage which is not yet repaired

`nrOfPictures` - the number of pictures in the ad

`postalCode` - the postal code for the location of the vehicle

`lastSeenOnline` - when the crawler saw this ad last online

In [1]:
## Import pandas & numpy libraries
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

- Most of the data in autos file is object, with a few in int64 type.
- vehicleType, gearbox, model, fuelType and notRepairedDamage are the column with a few missing values
- All of the columns contain non-null values

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

- Since the columns' names are in camelcase instead of Python's preferred snakecase, we should change the column's name based on the data dictionary to be more descriptive

In [7]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Basic data exploration

In [9]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-25 19:57:10  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

- Based on the statistics, we can see some of the columns that have mostly one value that are ideal for dropping: `seller`, `offer_type`, `ab_test`, `gear_box`, `unrepaired_damage`

In [10]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [11]:
autos['offer_type'].value_counts()


Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

- We can see that the column 'seller' & 'offer_type' have overwhelming number of 1 type of data over another, which shouls be taken into more investigation.

`price`, `object` are the columns with numeric data types but are stored as strings. 

### Clean the price and objects columns

In [12]:
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type   price  ab_test vehicle_type  registration_year   gear_box  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   
2    Angebot  $8,990     test    limousine               2009    manuell   

   power_PS   model   odometer  registration_month fuel_type       brand  \
0       158  andere  150,000km                   3       lpg     peugeot   
1       286     7er  150,000km                   6    benzin         bmw   
2       102    golf   70,000km                   7    benzin  volkswagen   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   
2              nein  2016-03-26 00:00:00               0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

In [13]:
# def clean_col(col):
#     col = col.strip()
#     col = col.replace('$', '')
#     return col

autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].str.replace(",", "")
autos['price'] = autos['price'].astype(int)

autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',', '')
autos['odometer'] = autos['odometer'].astype(int)

In [14]:
autos.rename({'odometer': 'odometer_km'}, axis = 1, inplace = True)

It looks like the `num_photos` column has 0 for every column. We'll drop this column + `seller` & `offer_type` column since all (nearly all) of the values are the same

In [15]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price  ab_test vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

    gear_box  power_PS   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [16]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  int64 
 5   ab_test             50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gear_box            47320 non-null  object
 9   power_PS            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer_km         50000 non-null  int64 
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

### Analyzing further odometer_km and price columns

In [17]:
autos['odometer_km'].unique().shape

(13,)

In [18]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [19]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [20]:
autos['odometer_km'].sort_index(ascending = False).head()

49999    150000
49998     40000
49997      5000
49996    150000
49995    100000
Name: odometer_km, dtype: int64

In [21]:
autos['price'].unique().shape

(2357,)

In [22]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [23]:
autos['price'].value_counts(ascending = False).head(100)

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
3100      133
4300      132
9999      131
2850      131
13500     130
Name: price, Length: 100, dtype: int64

In [24]:
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

## Exploring odometer and price columns

In [25]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

The value in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Besides, there are more high mileage than low mileage vehicles 

In [26]:
print(autos['price'].unique().shape)
print(autos['price'].describe())
print(autos['price'].value_counts().head(20))

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64


Again, the prices in this column seem rounded, however given there are 2357 unique values in the column, that may be just people's tendency to round prices on the site

There are 1,421 cars listed with $0 price - given that this is only 2% of the cars, we might consider removing these rows. The maximum price is 1 hundered million dollars, which seems a lot, let's look at the highest prices further

In [27]:
autos['price'].value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [28]:
autos['price'].value_counts().sort_index(ascending = True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

There are a number of listings with prices below \\$30\, including about 1,500 at \\$0. There are also a small number of listings with very high values, including 14 at around $1 millions

Given that eBay is an auction site, there could legitimately be items where the opening bid is \\$1. We will keep the \\$1 items, but remove anything above $350,000, since it seems that prices increase steadily to that number and then jump up to less than realistic numbers.

In [29]:
autos = autos[autos['price'].between(1, 351000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring the date columns

There are 5 columns representing date values. Some are created by the crawler, some came from the website itself

`date_crawled`: added by the crawler

`last_seen`: added by the crawler

`ad_created`: from the website

`registration_month`: from the website

`registration_year`: from the website

In [30]:
## Take a look at 5 columns
autos[['date_crawled', 'last_seen', 'ad_created', 'registration_month', 'registration_year']].head()

date_crawled            last_seen           ad_created  \
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00   
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00   
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00   
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00   
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00   

   registration_month  registration_year  
0                   3               2004  
1                   6               1997  
2                   7               2009  
3                   6               2007  
4                   7               2003

In [31]:
import datetime as dt

In [32]:
crawled_date = autos['date_crawled'].str[:10]

In [33]:
type(crawled_date)

pandas.core.series.Series

In [34]:
crawled_date.value_counts().sort_index()

2016-03-05    1230
2016-03-06     682
2016-03-07    1749
2016-03-08    1617
2016-03-09    1607
2016-03-10    1563
2016-03-11    1582
2016-03-12    1793
2016-03-13     761
2016-03-14    1775
2016-03-15    1665
2016-03-16    1438
2016-03-17    1536
2016-03-18     627
2016-03-19    1689
2016-03-20    1840
2016-03-21    1815
2016-03-22    1602
2016-03-23    1565
2016-03-24    1425
2016-03-25    1535
2016-03-26    1564
2016-03-27    1510
2016-03-28    1693
2016-03-29    1656
2016-03-30    1636
2016-03-31    1546
2016-04-01    1636
2016-04-02    1723
2016-04-03    1875
2016-04-04    1772
2016-04-05     636
2016-04-06     154
2016-04-07      68
Name: date_crawled, dtype: int64

As we can see, the crawled_date are distributed fairly equally, and the month recorded are only on March and April

In [35]:
crawled_date = crawled_date.str.replace('-', '')

In [36]:
crawled_date.head()

0    20160326
1    20160404
2    20160326
3    20160312
4    20160401
Name: date_crawled, dtype: object

In [37]:
crawled_date = crawled_date.astype(int)

In [38]:
type(crawled_date)

pandas.core.series.Series

In [39]:
crawled_date.describe()

count    4.856500e+04
mean     2.016033e+07
std      3.190128e+01
min      2.016030e+07
25%      2.016031e+07
50%      2.016032e+07
75%      2.016033e+07
max      2.016041e+07
Name: date_crawled, dtype: float64

In [40]:
crawled_date.value_counts(normalize = True, dropna= False)

20160403    0.038608
20160320    0.037887
20160321    0.037373
20160312    0.036920
20160314    0.036549
20160404    0.036487
20160307    0.036014
20160402    0.035478
20160328    0.034860
20160319    0.034778
20160315    0.034284
20160329    0.034099
20160330    0.033687
20160401    0.033687
20160308    0.033296
20160309    0.033090
20160322    0.032987
20160311    0.032575
20160323    0.032225
20160326    0.032204
20160310    0.032184
20160331    0.031834
20160317    0.031628
20160325    0.031607
20160327    0.031092
20160316    0.029610
20160324    0.029342
20160305    0.025327
20160313    0.015670
20160306    0.014043
20160405    0.013096
20160318    0.012911
20160406    0.003171
20160407    0.001400
Name: date_crawled, dtype: float64

80% of the `crawled_date` are distributed evenly, with most of the crawled date happened on March, 2016, while the rest occurred on April in the same year

In [41]:
crawled_date.value_counts().sort_index()

20160305    1230
20160306     682
20160307    1749
20160308    1617
20160309    1607
20160310    1563
20160311    1582
20160312    1793
20160313     761
20160314    1775
20160315    1665
20160316    1438
20160317    1536
20160318     627
20160319    1689
20160320    1840
20160321    1815
20160322    1602
20160323    1565
20160324    1425
20160325    1535
20160326    1564
20160327    1510
20160328    1693
20160329    1656
20160330    1636
20160331    1546
20160401    1636
20160402    1723
20160403    1875
20160404    1772
20160405     636
20160406     154
20160407      68
Name: date_crawled, dtype: int64

In [42]:
## Do the same thing with other 2 columns whose types are strings
last_seen_date = autos['last_seen'].str[:10]
ad_created_date = autos['ad_created'].str[:10]

In [43]:
ad_created_date.head()

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: ad_created, dtype: object

In [44]:
last_seen_date = last_seen_date.str.replace('-', '').astype(int)
ad_created_date = ad_created_date.str.replace('-', '').astype(int)

In [45]:
last_seen_date.value_counts().sort_index()

20160305       52
20160306      210
20160307      262
20160308      360
20160309      466
20160310      518
20160311      601
20160312     1155
20160313      432
20160314      612
20160315      771
20160316      799
20160317     1364
20160318      357
20160319      769
20160320     1003
20160321     1002
20160322     1038
20160323      900
20160324      960
20160325      933
20160326      816
20160327      760
20160328     1013
20160329     1085
20160330     1203
20160331     1155
20160401     1107
20160402     1210
20160403     1224
20160404     1189
20160405     6059
20160406    10772
20160407     6408
Name: last_seen, dtype: int64

In [46]:
ad_created_date.value_counts(normalize = True, dropna = False).sort_index().head(20)

20150611    0.000021
20150810    0.000021
20150909    0.000021
20151110    0.000021
20151205    0.000021
20151230    0.000021
20160103    0.000021
20160107    0.000021
20160110    0.000041
20160113    0.000021
20160114    0.000021
20160116    0.000021
20160122    0.000021
20160127    0.000062
20160129    0.000021
20160201    0.000021
20160202    0.000041
20160205    0.000041
20160207    0.000021
20160208    0.000021
Name: ad_created, dtype: float64

In [47]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The mean of the values in `registration_year` is 2004, which means most of the car have been registered during 2004. The maximum year is 9999, which is way too far in the future and make no sense so we need to further investigate this. The same thing happens with the minimum value - 1000 - which is before the car invented!

In [48]:
autos['registration_month'].describe()

count    48565.000000
mean         5.782251
std          3.685595
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

The mean of the values in `registration_month` is ~5, which means that most of the cars have been registered on May and June. While the maximum value indicates December, the minumim value said it was 0, which required more investigation

## Dealing with Incorrect Registration Year Data

In [49]:
sorted(autos['registration_year'].unique())

[1000,
 1001,
 1111,
 1800,
 1910,
 1927,
 1929,
 1931,
 1934,
 1937,
 1938,
 1939,
 1941,
 1943,
 1948,
 1950,
 1951,
 1952,
 1953,
 1954,
 1955,
 1956,
 1957,
 1958,
 1959,
 1960,
 1961,
 1962,
 1963,
 1964,
 1965,
 1966,
 1967,
 1968,
 1969,
 1970,
 1971,
 1972,
 1973,
 1974,
 1975,
 1976,
 1977,
 1978,
 1979,
 1980,
 1981,
 1982,
 1983,
 1984,
 1985,
 1986,
 1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013,
 2014,
 2015,
 2016,
 2017,
 2018,
 2019,
 2800,
 4100,
 4500,
 4800,
 5000,
 5911,
 6200,
 8888,
 9000,
 9999]

Based on this, the highest acceptable values can be **2019**, because from 2800 move on, the values are too high to make accurate estimation. The lowest acceptable value will **1910**, because its preceded value - 1800 - is way before the first car was invented

In [50]:
## Using `Series.between()` to select rows in a dataframe that fall within a value range for a column
autos = autos[autos['registration_year'].between(1910, 2019)]
autos['registration_year'].describe()

count    48545.000000
mean      2003.461510
std          7.566196
min       1910.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       2019.000000
Name: registration_year, dtype: float64

In [51]:
autos['registration_year'].value_counts(normalize = True).head(10)

2000    0.065012
2005    0.060480
1999    0.059677
2004    0.055680
2003    0.055598
2006    0.055001
2001    0.054300
2002    0.051210
1998    0.048676
2007    0.046905
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years

## Exploring Price by Brand

In [52]:
len(autos['brand'].unique())

40

In [53]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [54]:
autos['brand'].value_counts(normalize = True)

volkswagen        0.212813
opel              0.108641
bmw               0.108641
mercedes_benz     0.095787
audi              0.085858
ford              0.069667
renault           0.047894
peugeot           0.029457
fiat              0.025996
seat              0.018931
skoda             0.016068
nissan            0.015264
mazda             0.015223
smart             0.014296
citroen           0.014111
toyota            0.012586
hyundai           0.009950
sonstige_autos    0.009641
volvo             0.009043
mini              0.008611
mitsubishi        0.008178
honda             0.007993
kia               0.007107
alfa_romeo        0.006612
porsche           0.005912
suzuki            0.005891
chevrolet         0.005665
chrysler          0.003481
dacia             0.002657
daihatsu          0.002513
jeep              0.002204
subaru            0.002101
land_rover        0.002039
saab              0.001627
daewoo            0.001566
jaguar            0.001524
trabant           0.001380
r

German manufacturers represent 4 out of top 5 brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next 2 brands combined.

There are a lot of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings

In [55]:
brand_counts = autos['brand'].value_counts(normalize = True)
common_brand = brand_counts[brand_counts > 0.05].index

In [56]:
print(common_brand)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [57]:
brand_mean_prices = {}
for brand in common_brand:
    brand_only = autos[autos['brand'] == brand]
    mean_price = brand_only['price'].mean()
    brand_mean_prices[brand] = int(mean_price)
    
print(brand_mean_prices)

{'volkswagen': 5333, 'opel': 2941, 'bmw': 8261, 'mercedes_benz': 8526, 'audi': 9212, 'ford': 3728}


Of the top 5 brands, there is a distinct gap:
- Audi, BMW, Mercedes_benz are the most expensive
- Ford & Opel are the lest expensive
- Volkswagen is in between - this may explains its popularity (which is the best of both world)

In [58]:
top_20_car_brand = ['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault', 'peugeot', 
                    'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart', 'citroen', 'toyota', 'huyndai',
                    'sonstige_autos', 'volve', 'mini']
mean_price_brand = {}
for row in top_20_car_brand:
    brand_only = autos[autos['brand'] == row]
    mean_price = brand_only['price'].mean()
    mean_price_brand[row] = mean_price
    
print(sorted(mean_price_brand.items())) ## Sorting alphabetically

[('audi', 9212.9306621881), ('bmw', 8261.382442169132), ('citroen', 3756.07299270073), ('fiat', 2793.8700475435817), ('ford', 3728.4121821407452), ('huyndai', nan), ('mazda', 4059.059539918809), ('mercedes_benz', 8526.623225806452), ('mini', 10541.566985645934), ('nissan', 4669.3859649122805), ('opel', 2941.4664391353813), ('peugeot', 3065.611888111888), ('renault', 2431.195698924731), ('seat', 4320.168661588684), ('skoda', 6353.544871794872), ('smart', 3518.102305475504), ('sonstige_autos', 12208.673076923076), ('toyota', 5148.0032733224225), ('volkswagen', 5333.1962055948115), ('volve', nan)]


## Sorting Aggregate data in a Dataframe

In [59]:
# Calculating mean mileage and mean price for each of the top brand
top_mean_lineage = {}
for brand in common_brand:
    brand_only = autos[autos['brand'] == brand]
    mean_lineage = brand_only['odometer_km'].mean()
    top_mean_lineage[brand] = mean_lineage
print(top_mean_lineage)

{'volkswagen': 128955.570612719, 'opel': 129452.02882062951, 'bmw': 132682.97307546454, 'mercedes_benz': 130848.3870967742, 'audi': 129492.56238003839, 'ford': 124349.49733885274}


In [60]:
# Convert both dictionary into objetcs
mileage_series = pd.Series(top_mean_lineage)
price_series = pd.Series(brand_mean_prices)
print(mileage_series, price_series)

volkswagen       128955.570613
opel             129452.028821
bmw              132682.973075
mercedes_benz    130848.387097
audi             129492.562380
ford             124349.497339
dtype: float64 volkswagen       5333
opel             2941
bmw              8261
mercedes_benz    8526
audi             9212
ford             3728
dtype: int64


In [61]:
# Create a dataframe from the first series object
df = pd.DataFrame(mileage_series, columns = ['mean_mileage'])

# Assign other series a new column
df.insert(1, 'mean_price', price_series)
print(df)

                mean_mileage  mean_price
volkswagen     128955.570613        5333
opel           129452.028821        2941
bmw            132682.973075        8261
mercedes_benz  130848.387097        8526
audi           129492.562380        9212
ford           124349.497339        3728


From the chart, we can see that:
- Audi, Mercedes_benz, BMW are the most expensive and covers the most mileage in the top brand
- However, the difference between the average mileage from the most expensive car to the least one are not pronounced. Ford car with the price of approximately 3728 almost covers 125,000 km, in comparison with Audi (cost ~ 9300), which covers roughly 129,000km 
- Therefore, we can conclude that there's no specific relatiship between mileage and the mean price. But by comparing mean price with the average lineage of a specific car, customers can make more approapriate choices 